application/tests/org.openjdk.jmc.jolokia.test/src/test/java/org/openjdk/jmc/jolokia/JolokiaTest.java

```diff
@@ -0,0 +1,163 @@
+import org.junit.Assert;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.openjdk.jmc.common.IDescribable;
+import org.openjdk.jmc.jolokia.preferences.PreferenceConstants;
+import org.openjdk.jmc.rjmx.IConnectionDescriptor;
+import org.openjdk.jmc.rjmx.IServerDescriptor;
+import org.openjdk.jmc.rjmx.descriptorprovider.IDescriptorListener;
+
+/**
+ * I test that JMX connections done with JmcJolokiaJmxConnectionProvider are functional 
+ */
+@SuppressWarnings("restriction")
+public class JolokiaTest {
+
+	static String jolokiaUrl;
+
+	private static Set<String> unsafeAttributes = new HashSet<>(
+			Arrays.asList("BootClassPath", "UsageThreshold", "UsageThresholdExceeded", "UsageThresholdCount",
+					"CollectionUsageThreshold", "CollectionUsageThresholdExceeded", "CollectionUsageThresholdCount"));
+
+	private static JolokiaDiscoveryListener discoveryListener;
+
+	private static MBeanServerConnection jolokiaConnection;
+
+	@BeforeClass
+	public static void startServer() throws Exception {
+		// wait for Jolokia to be ready before commencing tests
+		Awaitility.await().atMost(Duration.ofSeconds(15))//Note: hard code property to avoid module dependency on agent
+				.until(() -> (jolokiaUrl = System.getProperty("jolokia.agent")) != null);
+		discoveryListener = new JolokiaDiscoveryListener();
+		jolokiaConnection=getJolokiaMBeanConnector();
+		
+
+	}
+
+	@Test
+	public void testReadAttributesOverJolokia() throws MalformedURLException, IOException, OperationsException,
+			IntrospectionException, AttributeNotFoundException, ReflectionException, MBeanException {
+		for (ObjectName objectName : jolokiaConnection.queryNames(null, null)) {
+			for (MBeanAttributeInfo attributeInfo : getJolokiaMBeanConnector().getMBeanInfo(objectName).getAttributes()) {
+				if (!unsafeAttributes.contains(attributeInfo.getName())) {
+					System.out.println("Getting attribute " + objectName + "/" + attributeInfo.getName());
+					getJolokiaMBeanConnector().getAttribute(objectName, attributeInfo.getName());
+				}
+			}
+		}
+
+	}
+	
+	@Test
+	public void testExecuteOperation() throws InstanceNotFoundException, MalformedObjectNameException, MBeanException, ReflectionException, MalformedURLException, IOException {
+		jolokiaConnection.invoke(new ObjectName("java.lang:type=Memory"), "gc", new Object[0], new String[0]);
+	}
+	
+	@Test
+	public void testWriteAttribute() throws InstanceNotFoundException, AttributeNotFoundException, InvalidAttributeValueException, MalformedObjectNameException, MBeanException, ReflectionException, MalformedURLException, IOException {
+		ObjectName objectName = new ObjectName("jolokia:type=Config");
+		String attribute = "Debug";
+		jolokiaConnection.setAttribute(objectName, new Attribute(attribute, true));
+		Assert.assertEquals(true, jolokiaConnection.getAttribute(objectName, attribute));
+		
+	}
+
+	private static MBeanServerConnection getJolokiaMBeanConnector() throws IOException, MalformedURLException {
```

Comment:
Testing discovery may pose for CI environments (?). We can consider making it manual if networking is an issue